File: melonet-backend/internal/domain/api/voice_cover.go
package api

import "time"

type VoiceArtistResponse struct {
	ID          int64  `json:"id"`
	Slug        string `json:"slug"`
	DisplayName string `json:"display_name"`
	AvatarURL   string `json:"avatar_url"`
	Enabled     bool   `json:"enabled"`
}

type VoiceCoverResponse struct {
	ID               int64      `json:"id"`
	SourceSongID     string     `json:"source_song_id"`
	TargetArtistID   int64      `json:"target_artist_id"`
	TargetArtistSlug string     `json:"target_artist_slug"`
	TargetArtistName string     `json:"target_artist_name"`
	Status           string     `json:"status"`
	AudioURL         string     `json:"audio_url,omitempty"`
	CoverURL         string     `json:"cover_url"`
	SourceTitle      string     `json:"source_title"`
	SourceArtist     string     `json:"source_artist"`
	Error            string     `json:"error,omitempty"`
	ProgressPct      int        `json:"progress_pct"`
	ProgressStage    string     `json:"progress_stage"`
	EtaSeconds       int        `json:"eta_seconds"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
	ReadyAt          *time.Time `json:"ready_at,omitempty"`
}

type CreateVoiceCoverRequest struct {
	SourceSongID     string `json:"source_song_id" binding:"required"`
	TargetArtistSlug string `json:"target_artist_slug" binding:"required"`
}

File: melonet-android/app/src/main/java/com/melonet/app/data/model/VoiceCover.kt
package com.melonet.app.data.model

import androidx.compose.runtime.Immutable

@Immutable
data class VoiceArtist(
    val id: Long,
    val slug: String,
    val displayName: String,
    val avatarUrl: String,
    val enabled: Boolean,
)

@Immutable
data class VoiceCover(
    val id: Long,
    val sourceSongId: String,
    val targetArtistId: Long,
    val targetArtistSlug: String,
    val targetArtistName: String,
    val status: String,
    val audioUrl: String,
    val coverUrl: String,
    val sourceTitle: String,
    val sourceArtist: String,
    val error: String,
    val progressPct: Int = 0,
    val progressStage: String = "",
    val etaSeconds: Int = 0,
) {
    val isReady: Boolean get() = status == "ready" && audioUrl.isNotBlank()
    val isFailed: Boolean get() = status == "failed"
    val isInProgress: Boolean get() = status == "pending" || status == "processing"

    fun toSong(): Song = Song(
        id = "vc-$id",
        title = sourceTitle.ifBlank { "Voice Cover #$id" },
        artistName = targetArtistName.ifBlank { sourceArtist },
        coverUrl = coverUrl,
        audioUrl = audioUrl,
        category = "voice_cover",
        lyrics = "",
        durationSec = 0,
        genre = targetArtistName,
    )
}
